src/utils/Queue/Queue.js

```diff
@@ -0,0 +1,65 @@
+const Emitter = require('events').EventEmitter;
+
+class Queue extends Emitter {
+  constructor(options = { autoProcess: true }) {
+    super();
+    this.jobs = [];
+    this.state = {
+      isProcessing: false,
+    };
+    this.autoProcess = options.autoProcess || true;
+
+    if (this.autoProcess) {
+      this.startAutoProcessing();
+    }
+  }
+
+  getSize() {
+    return this.jobs.length;
+  }
+
+  enqueueJob(job) {
+    this.jobs.push(job);
+    this.emit('enqueued', job.id);
+  }
+
+  dequeueJob() {
+    const job = this.jobs.shift();
+    this.emit('dequeued', job);
+    return job;
+  }
+
+  async processNext() {
+    const job = this.dequeueJob();
+    if (job) {
+      await this.processJob(job);
+    }
+  }
+
+  async processJob(job) {
+    this.state.isProcessing = true;
+    const result = await job.fn();
+    this.state.isProcessing = false;
+    this.emit('processed', { result, job });
+    return result;
+  }
+
+  startAutoProcessing() {
+    const self = this;
+
+    const processOnEnqueuedEvent = async () => {
+      await self.processNext();
+    };
+    const processOnProcessedEvent = async () => {
+      await self.processNext();
+      if (!self.getSize()) {
+        this.once('enqueued', processOnEnqueuedEvent);
+      }
+    };
+
+    this.on('processed', processOnProcessedEvent);
```

Comment:
Errors from jobs are not handled since you use syncronus event emitter with async tasks